### ISSUE TITLE
Feature Request: Cancel pending metric queries when new ones are issued 

### ISSUE THREAD

**Issue by HallsHalls** [grafana/grafana#2501]
I have thousands of millions of data to query ,so too hard to get result by frequently operating, like that

![2015-08-13 08 33 24](https://cloud.githubusercontent.com/assets/9986416/9240626/0c2f6b54-419b-11e5-9654-0914a60cd517.png)

I just try to refresh or do some same operating frequently. The question is how can I to cancel the previous query request, I just want to the latest response result .

By the way , I want to add some judge( 30ms~40ms same operator do not send request ) from operate to stop the frequently request.

**Comment by torkelo** [comment]
Not sure, Grafana is built for time series databases, a time series query usually take around 30-100ms no matter what time range (given you have good rollup retention periods). The extreme query might take 1second but that is very rare. 

What time series database are you using that is not responding?

**Comment by HallsHalls** [comment]
Our team use OpenTSDB database. And we try our best to reduce query time. but some extreme query ,especially on dashboard which have many graph query .
like that ,there have twice same request, but the first request are pending and I want to stop them even kill them ,the second operator response result must wait all the first time request over, that user will be crazy or  try to do the same operate(like refresh), and wait forever.
![123](https://cloud.githubusercontent.com/assets/9986416/9268957/d6f64dca-4296-11e5-8a74-190e29bb1862.png)

**Comment by HallsHalls** [comment]
@torkelo

**Comment by utkarshcmu** [comment]
+1

**Comment by wamdam** [comment]
We have the same problem, some queries are very expensive.

Grafana should cancel all pending queries when the dashboard is reloaded/reconfigured by cancelling the http requests so that the backend database can detect cancelled queries. In addition, if grafana proxies the request to the database, it should also close the proxy connection.
According to http://stackoverflow.com/questions/13928057/how-to-cancel-an-http-request-in-angularjs this should be possible via the timeout parameter.

```
var canceler = $q.defer();
$http.get('/someUrl', {timeout: canceler.promise}).success(successCallback);
// later...
canceler.resolve();  // Aborts the $http request if it isn't finished.
```

We have our own database adapter so it'd be easy for us to implement this in the database web server.

Additionally it *seems* like all returned results (also the old ones from the previous queries) will be rendered as graphs, which is a bit weird.

**Comment by torkelo** [comment]
Rendered as graphs?

**Comment by wamdam** [comment]
What I meant was that every data graphana receives (even outdated queries) are rendered when they are received.

**Comment by grobie** [comment]
:+1:

I don't think that "a time series query usually take around 30-100ms no matter what time range" is a general feature of time series databases.

**Comment by jifwin** [comment]
+1

**Comment by evanbreznyik** [comment]
+1

**Comment by smalik03** [comment]
+1

**Comment by grobie** [comment]
@torkelo Not stopping in-flight requests when starting new ones continues to cause a lot of issues for us, especially because we use the proxy mode. Any objections to accept a solution based on what @wamdam already suggested?

**Comment by torkelo** [comment]
Agree this should be fixed. No objections, a PR with that solution would likely be accepted

**Comment by torkelo** [comment]
this was implemented in 3.1
